Add guide how to access Scalar products on Kubernetes (#196)
scalar-labs/scalar-kubernetes: docs/AccessScalarProducts.md

> # Expose ScalarDB or ScalarDL in a Kubernetes cluster environment

Reviewer: ```suggestion
# Make ScalarDB or ScalarDL deployed in a Kubernetes cluster environment available from applications
```

Reviewer 2: Good catch. LGTM!

> This document explains how to expose ScalarDB or ScalarDL in a Kubernetes cluster environment. To expose ScalarDB or ScalarDL, you can use Scalar Envoy via a Kubernetes service resource named `<HELM_RELEASE_NAME>-envoy`. You can use `<HELM_RELEASE_NAME>-envoy` in several ways, such as:

Reviewer: I feel if we say `expose`, it should be like `expose ScalarDB or ScalarDL endpoints/APIs`, but it still sounds a bit unnatural (indirect?) to me. Also, `in` makes things confusing since we actually expose endpoints outside the K8s cluster.

How about saying this?

```suggestion
This document explains how to make ScalarDB or ScalarDL deployed in a Kubernetes cluster environment available from applications. To expose ScalarDB or ScalarDL, you can use Scalar Envoy via a Kubernetes service resource named `<HELM_RELEASE_NAME>-envoy`. You can use `<HELM_RELEASE_NAME>-envoy` in several ways, such as:
```
@josh-wong What do you think?

Reviewer 2: Thanks for the explanation. LGTM!

Author: Thank you for your suggestion!
It looks good to me. I applied it.

When using the Kubernetes service resource, you must set the above FQDN in the properties file for the application (client) as follows:

* **Client properties file for ScalarDB Server**
    ```properties
    scalar.db.contact_points=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.db.contact_port=60051
    scalar.db.storage=grpc
    scalar.db.transaction_manager=grpc
    ```
* **Client properties file for ScalarDL Ledger**
    ```properties
    scalar.dl.client.server.host=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.dl.ledger.server.port=50051
    scalar.dl.ledger.server.privileged_port=50052
    ```
* **Client properties file for ScalarDL Ledger with ScalarDL Auditor mode enabled**
    ```properties
    # Ledger
    scalar.dl.client.server.host=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.dl.ledger.server.port=50051
    scalar.dl.ledger.server.privileged_port=50052
    
    # Auditor
    scalar.dl.client.auditor.enabled=true
    scalar.dl.client.auditor.host=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.dl.auditor.server.port=40051
    scalar.dl.auditor.server.privileged_port=40052
    ```

> ## Expose ScalarDB or ScalarDL via load balancers to run application (client) requests from outside the Kubernetes cluster

Reviewer: It depends on how we say it, but I think we should use the same wording and sentence structure as the corresponding sentence (`Run application (client) requests to  ...`) above.

Reviewer 2: I see. Following the same structure as the previous heading, how about we use this?:

```suggestion
## Run application (client) requests to ScalarDB or ScalarDL via load balancers from outside the Kubernetes cluster
```

Author: Thank you for your suggestion!
It looks good to me. I applied it.

> # Access Scalar products on a Kubernetes environment

Reviewer 2: I changed "Scalar products" to "ScalarDB and ScalarDL" throughout this document to be more specific. I also incorporated some suggestions from [this comment within the doc](https://github.com/scalar-labs/scalar-kubernetes/pull/196/files#r1128986435).

If either of those are incorrect assumptions, let's discuss!

```suggestion
# Expose ScalarDB or ScalarDL in a Kubernetes cluster environment
```

Author: Thank you for incorporating our comments and your suggestions!
The suggested titles look good to me!

> This document explains how to access Scalar products on a Kubernetes environment. On a Kubernetes environment, you can access Scalar products via Scalar Envoy. To access this Scalar Envoy via service resource of Kubernetes named as `<helm release name>-envoy`. There are several ways to access this `<helm release name>-envoy` service resource as follows.

Reviewer: I think `access` might be too blurry here.
We probably mean requesting transactions to Scalar products, right?
I first thought it was for doing `kubetl get pods` to Scalar products.

Author: @feeblefakie  
Thank you for your review!

> I think `access` might be too blurry here.
> We probably mean requesting transactions to Scalar products, right?

Yes. That's right.
I agree `access` is blurry.

I think there are several options instead of `access Scalar product` as follows.

* run client requests to Scalar products
* execute client requests to Scalar products
* accept client requests for Scalar products
* listen for client requests on Scalar products
* expose Scalar products to run/execute client requests

And, in my opinion, I think we can update each title/section title as follows. (Also, I will update all statements in this document based on each title.)

* h1 (title): Expose Scalar products on a Kubernetes environment
* h2 (section title): Run client requests to Scalar products via service resources directly from inside of the Kubernetes cluster
  * Note: In this case, we don't need additional operations other than deploying to expose Scalar products. Scalar Envoy can accept client requests from inside the Kubernetes cluster by default.
* h2 (section title): Expose Scalar products via Load Balancers to run client requests from outside of the Kubernetes cluster
* h2 (section title):  Run client requests to Scalar products from a bastion server (For testing purposes only / Not recommended in the production environment)

The most popular use case in the production environment is `Expose Scalar products via Load Balancers`, so I think it is better to use `expose` for the document title, too.

What do you think?

> This document explains how to access Scalar products on a Kubernetes environment. On a Kubernetes environment, you can access Scalar products via Scalar Envoy. To access this Scalar Envoy via service resource of Kubernetes named as `<helm release name>-envoy`. There are several ways to access this `<helm release name>-envoy` service resource as follows.

Reviewer 2: ```suggestion
This document explains how to expose ScalarDB or ScalarDL in a Kubernetes cluster environment. To expose ScalarDB or ScalarDL, you can use Scalar Envoy via a Kubernetes service resource named `<HELM_RELEASE_NAME>-envoy`. You can use `<HELM_RELEASE_NAME>-envoy` in several ways, such as:
```

> * Access `<helm release name>-envoy` directly from **inside** of the same Kubernetes cluster as Scalar products.
> * Access `<helm release name>-envoy` via a Load Balancer from **outside** of the Kubernetes cluster.
> * Access `<helm release name>-envoy` from a bastion server using the `kubectl port-forward` command (testing purpose only).
> 
> Also, the resource name `<helm release name>-envoy` is decided based on the helm release name. You can see the helm release name using the `helm list` command.

Reviewer 2: ```suggestion
* Directly from inside the same Kubernetes cluster as ScalarDB or ScalarDL.
* Via a load balancer from outside the Kubernetes cluster.
* From a bastion server by using the `kubectl port-forward` command (for testing purposes only).

The resource name `<HELM_RELEASE_NAME>-envoy` is decided based on the helm release name. You can see the helm release name by running the `helm list` command.
```

```console
$ helm list -n ns-scalar
NAME                    NAMESPACE       REVISION        UPDATED                                 STATUS                                                       CHART                    APP VERSION
scalardb                ns-scalar       1               2023-02-09 19:31:40.527130674 +0900 JST deployed                                                     scalardb-2.5.0           3.8.0
scalardl-auditor        ns-scalar       1               2023-02-09 19:32:03.008986045 +0900 JST deployed                                                     scalardl-audit-2.5.1     3.7.1
scalardl-ledger         ns-scalar       1               2023-02-09 19:31:53.459548418 +0900 JST deployed                                                     scalardl-4.5.1           3.7.1
```

> You can also see the envoy service name `<helm release name>-envoy` directly using the `kubectl get service` command.

Reviewer 2: ```suggestion
You can also see the envoy service name `<HELM_RELEASE_NAME>-envoy` by running the `kubectl get service` command.
```

scalardb-headless                ClusterIP      None             <none>        60051/TCP                         2m2s
scalardb-metrics                 ClusterIP      10.111.213.194   <none>        8080/TCP                          2m2s
scalardl-auditor-envoy           LoadBalancer   10.111.141.43    <pending>     40051:31553/TCP,40052:31171/TCP   99s
scalardl-auditor-envoy-metrics   ClusterIP      10.104.245.188   <none>        9001/TCP                          99s
scalardl-auditor-headless        ClusterIP      None             <none>        40051/TCP,40053/TCP,40052/TCP     99s
scalardl-auditor-metrics         ClusterIP      10.105.119.158   <none>        8080/TCP                          99s
scalardl-ledger-envoy            LoadBalancer   10.96.239.167    <pending>     50051:32714/TCP,50052:30857/TCP   109s
scalardl-ledger-envoy-metrics    ClusterIP      10.97.204.18     <none>        9001/TCP                          109s
scalardl-ledger-headless         ClusterIP      None             <none>        50051/TCP,50053/TCP,50052/TCP     109s
scalardl-ledger-metrics          ClusterIP      10.104.216.189   <none>        8080/TCP                          109s
```

> ## Access Scalar products via service resources directly from inside of the Kubernetes cluster

Reviewer 2: ```suggestion
## Run application (client) requests to ScalarDB or ScalarDL via service resources directly from inside the same Kubernetes cluster
```

> If you deploy your application (client) on the same Kubernetes cluster as Scalar products (e.g., you deploy your application on another node group/pool on the same Kubernetes cluster), the application can access Scalar products using service resources of Kubernetes. The format of the service resource name (FQDN) is `<helm release name>-envoy.<namespace>.svc.cluster.local`.
> 
> * Example (if you deploy Scalar products on the `ns-scalar` namespace)
>   * ScalarDB Server
>     ```console
>     scalardb-envoy.ns-scalar.svc.cluster.local
>     ```
>   * ScalarDL Ledger
>     ```console
>     scalardl-ledger-envoy.ns-scalar.svc.cluster.local
>     ```
>   * ScalarDL Auditor
>     ```console
>     scalardl-auditor-envoy.ns-scalar.svc.cluster.local
>     ```

Reviewer 2: ```suggestion
If you deploy your application (client) in the same Kubernetes cluster as ScalarDB or ScalarDL (for example, if you deploy your application [client] on another node group or pool in the same Kubernetes cluster), the application can access ScalarDB or ScalarDL by using Kubernetes service resources. The format of the service resource name (FQDN) is `<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local`.

The following are examples of ScalarDB and Scalar DL deployments on the `ns-scalar` namespace:

* **ScalarDB Server**
    ```console
    scalardb-envoy.ns-scalar.svc.cluster.local
    ```
* **ScalarDL Ledger**
    ```console
    scalardl-ledger-envoy.ns-scalar.svc.cluster.local
    ```
* **ScalarDL Auditor**
    ```console
    scalardl-auditor-envoy.ns-scalar.svc.cluster.local
    ```
```

> If you use the service resource of Kubernetes, you need to set the above FQDN in the properties file for the application (client) as follows.
> 
> * Client properties file for ScalarDB Server
>   ```properties
>   scalar.db.contact_points=<helm release name>-envoy.<namespace>.svc.cluster.local
>   scalar.db.contact_port=60051
>   scalar.db.storage=grpc
>   scalar.db.transaction_manager=grpc
>   ```
> * Client properties file for ScalarDL Ledger
>   ```properties
>   scalar.dl.client.server.host=<helm release name>-envoy.<namespace>.svc.cluster.local
>   scalar.dl.ledger.server.port=50051
>   scalar.dl.ledger.server.privileged_port=50052
>   ```
> * Client properties file for ScalarDL Ledger and ScalarDL Auditor (Auditor mode)
>   ```properties
>   # Ledger
>   scalar.dl.client.server.host=<helm release name>-envoy.<namespace>.svc.cluster.local
>   scalar.dl.ledger.server.port=50051
>   scalar.dl.ledger.server.privileged_port=50052
>   
>   # Auditor
>   scalar.dl.client.auditor.enabled=true
>   scalar.dl.client.auditor.host=<helm release name>-envoy.<namespace>.svc.cluster.local
>   scalar.dl.auditor.server.port=40051
>   scalar.dl.auditor.server.privileged_port=40052
>   ```

Reviewer 2: ```suggestion
When using the Kubernetes service resource, you must set the above FQDN in the properties file for the application (client) as follows:

* **Client properties file for ScalarDB Server**
    ```properties
    scalar.db.contact_points=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.db.contact_port=60051
    scalar.db.storage=grpc
    scalar.db.transaction_manager=grpc
    ```
* **Client properties file for ScalarDL Ledger**
    ```properties
    scalar.dl.client.server.host=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.dl.ledger.server.port=50051
    scalar.dl.ledger.server.privileged_port=50052
    ```
* **Client properties file for ScalarDL Ledger with ScalarDL Auditor mode enabled**
    ```properties
    # Ledger
    scalar.dl.client.server.host=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.dl.ledger.server.port=50051
    scalar.dl.ledger.server.privileged_port=50052
    
    # Auditor
    scalar.dl.client.auditor.enabled=true
    scalar.dl.client.auditor.host=<HELM_RELEASE_NAME>-envoy.<NAMESPACE>.svc.cluster.local
    scalar.dl.auditor.server.port=40051
    scalar.dl.auditor.server.privileged_port=40052
    ```
```

> ## Access Scalar products via Load Balancers from outside of the Kubernetes cluster

Reviewer 2: ```suggestion
## Expose ScalarDB or ScalarDL via load balancers to run application (client) requests from outside the Kubernetes cluster
```

> If you deploy your application (client) on another environment than the Kubernetes cluster for Scalar products (e.g., you deploy your application on another  Kubernetes cluster, container platform, or server), the application can access Scalar products using a Load Balancer provided by each cloud service.

Reviewer 2: ```suggestion
If you deploy your application (client) in an environment outside the Kubernetes cluster for ScalarDB or ScalarDL (for example, if you deploy your application [client] on another Kubernetes cluster, container platform, or server), the application can access ScalarDB or ScalarDL by using a load balancer that each cloud service provides.
```

> You can create the Load Balancer to set `envoy.service.type` to `LoadBalancer` in your custom values file. When you create it, you can access Scalar Envoy (A service resource of Kubernetes) via created Load Balancer. You can also configure the Load Balancer configurations using annotations. Please refer to the following document for more details on how to configure your custom values file.
> 
> * [Service configurations](https://github.com/scalar-labs/helm-charts/blob/main/docs/configure-custom-values-envoy.md#service-configurations)
> 
> If you use the Load Balancer, you need to set the FQDN or IP address of the Load Balancer in the properties file for the application (client) as follows.

Reviewer 2: ```suggestion
You can create a load balancer by setting `envoy.service.type` to `LoadBalancer` in your custom values file. After configuring the custom values file, you can use Scalar Envoy as a Kubernetes service resource through the load balancer that you created. You can also configure the load balancer configurations by using annotations.

For more details on how to configure your custom values file, see [Service configurations](https://github.com/scalar-labs/helm-charts/blob/main/docs/configure-custom-values-envoy.md#service-configurations).

When using a load balancer, you must set the FQDN or IP address of the load balancer in the properties file for the application (client) as follows.
```

Author: Sorry for a very minor thing, but strictly the Scalar Envoy doesn't equal Kubernetes service. The request flow is the following.
```
[Client] ---> [Load balancer] ---> [Kubernetes service (k8s service)] ---> [Scalar Envoy (k8s pod)] ---> [ScalarDB/DL (k8s pod)]
```

So, I think the following explanation is better.
```
After configuring the custom values file, you can use Scalar Envoy through Kubernetes service resource through the load balancer that you created.
```

However, this sentence uses several `through` like `use AAA through BBB through CCC`.
Is this OK from the English/documentation perspective?
If no, do you have any idea to represent the above request flow in the document?

Author: > I think we should avoid using "through" more than once in this sentence. How about we change the last "through" to "by using"?

Looks good to me! Thank you!
I will refer to this when writing documentation in the future.

> * Client properties file for ScalarDB Server
>   ```properties
>   scalar.db.contact_points=<FQDN or IP address of Load Balancer>
>   scalar.db.contact_port=60051
>   scalar.db.storage=grpc
>   scalar.db.transaction_manager=grpc
>   ```
> * Client properties file for ScalarDL Ledger
>   ```properties
>   scalar.dl.client.server.host=<FQDN or IP address of Load Balancer>
>   scalar.dl.ledger.server.port=50051
>   scalar.dl.ledger.server.privileged_port=50052
>   ```
> * Client properties file for ScalarDL Ledger and ScalarDL Auditor (Auditor mode)
>   ```properties
>   # Ledger
>   scalar.dl.client.server.host=<FQDN or IP address of Load Balancer>
>   scalar.dl.ledger.server.port=50051
>   scalar.dl.ledger.server.privileged_port=50052
>   
>   # Auditor
>   scalar.dl.client.auditor.enabled=true
>   scalar.dl.client.auditor.host=<FQDN or IP address of Load Balancer>
>   scalar.dl.auditor.server.port=40051
>   scalar.dl.auditor.server.privileged_port=40052
>   ```

Reviewer 2: ```suggestion
* **Client properties file for ScalarDB Server**
    ```properties
    scalar.db.contact_points=<LOAD_BALANCER_FQDN_OR_IP_ADDRESS>
    scalar.db.contact_port=60051
    scalar.db.storage=grpc
    scalar.db.transaction_manager=grpc
    ```
* **Client properties file for ScalarDL Ledger**
    ```properties
    scalar.dl.client.server.host=<LOAD_BALANCER_FQDN_OR_IP_ADDRESS>
    scalar.dl.ledger.server.port=50051
    scalar.dl.ledger.server.privileged_port=50052
    ```
* **Client properties file for ScalarDL Ledger with ScalarDL Auditor mode enabled**
    ```properties
    # Ledger
    scalar.dl.client.server.host=<LOAD_BALANCER_FQDN_OR_IP_ADDRESS>
    scalar.dl.ledger.server.port=50051
    scalar.dl.ledger.server.privileged_port=50052
    
    # Auditor
    scalar.dl.client.auditor.enabled=true
    scalar.dl.client.auditor.host=<LOAD_BALANCER_FQDN_OR_IP_ADDRESS>
    scalar.dl.auditor.server.port=40051
    scalar.dl.auditor.server.privileged_port=40052
    ```
```

> The concrete implementation of the Load Balancer and access method depend on the Kubernetes cluster. If you use a managed Kubernetes cluster, please refer to the following cloud provider's official documents for more details.
> 
> ### EKS
> 
> * [Network load balancing on Amazon EKS](https://docs.aws.amazon.com/eks/latest/userguide/network-load-balancing.html)
> 
> ### AKS
> 
> * [Use a public standard load balancer in Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/load-balancer-standard)
> * [Use an internal load balancer with Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/internal-lb)

Reviewer 2: ```suggestion
The concrete implementation of the load balancer and access method depend on the Kubernetes cluster. If you are using a managed Kubernetes cluster, see the following official documentation based on your cloud service provider:

* **Amazon Elastic Kubernetes Service (EKS)**
  * [Network load balancing on Amazon EKS](https://docs.aws.amazon.com/eks/latest/userguide/network-load-balancing.html)
* **Azure Kubernetes Service (AKS)**
  * [Use a public standard load balancer in Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/load-balancer-standard)
  * [Use an internal load balancer with Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/internal-lb)
```

> ## Access Scalar products from a bastion server (For testing purposes only / Not recommended in the production environment)

Reviewer 2: ```suggestion
## Run client requests to ScalarDB or ScalarDL from a bastion server (for testing purposes only; not recommended in a production environment)
```

> You can access Scalar products from a bastion server using the `kubectl port-forward` command. If you create a ScalarDL Auditor mode environment, you need to access two Kubernetes clusters from one bastion server to access ScalarDL this way.

Reviewer 2: ```suggestion
You can run client requests to ScalarDB or ScalarDL from a bastion server by running the `kubectl port-forward` command. If you create a ScalarDL Auditor mode environment, however, you must run two Kubernetes clusters from one bastion server to access ScalarDL.
```

Reviewer 2: Understood. Your explanation makes perfect sense to me! 

How about we use the following?

```suggestion
You can run client requests to ScalarDB or ScalarDL from a bastion server by running the `kubectl port-forward` command. If you create a ScalarDL Auditor mode environment, however, you must run two `kubectl port-forward` commands with different kubeconfig files from one bastion server to access two Kubernetes clusters.
```

Author: Looks good to me! 
Thank you for updating your suggestion.
I apply this suggestion.

> 1. (ScalarDL Auditor mode only) In the bastion server for ScalarDL Ledger, configure (add) the kubeconfig to access the Kubernetes cluster of ScalarDL Auditor.
> 
>    Please refer to the following document for more details on how to configure the kubeconfig of each managed Kubernetes cluster.
> 
>    * [Configure kubeconfig](./CreateBastionServer.md#configure-kubeconfig)
> 
> 1. Port forwarding to each service from the bastion server.
>    * ScalarDB Server
>      ```console
>      kubectl port-forward -n <namespace> svc/<release name>-envoy 60051:60051
>      ```
>    * ScalarDL Ledger
>      ```console
>      kubectl --context <context of k8s for Ledger> port-forward -n <namespace> svc/<release name>-envoy 50051:50051
>      kubectl --context <context of k8s for Ledger> port-forward -n <namespace> svc/<release name>-envoy 50052:50052
>      ```
>    * ScalarDL Auditor
>      ```console
>      kubectl --context <context of k8s for Auditor> port-forward -n <namespace> svc/<release name>-envoy 40051:40051
>      kubectl --context <context of k8s for Auditor> port-forward -n <namespace> svc/<release name>-envoy 40052:40052
>      ```
> 
> 1. Configure properties file to access each Scalar product via `localhost`. 
>    * Client properties file for ScalarDB Server
>      ```properties
>      scalar.db.contact_points=localhost
>      scalar.db.contact_port=60051
>      scalar.db.storage=grpc
>      scalar.db.transaction_manager=grpc
>      ```
>    * Client properties file for ScalarDL Ledger
>      ```properties
>      scalar.dl.client.server.host=localhost
>      scalar.dl.ledger.server.port=50051
>      scalar.dl.ledger.server.privileged_port=50052
>      ```
>    * Client properties file for ScalarDL Ledger and ScalarDL Auditor (Auditor mode)
>      ```properties
>      # Ledger
>      scalar.dl.client.server.host=localhost
>      scalar.dl.ledger.server.port=50051
>      scalar.dl.ledger.server.privileged_port=50052
>      
>      # Auditor
>      scalar.dl.client.auditor.enabled=true
>      scalar.dl.client.auditor.host=localhost
>      scalar.dl.auditor.server.port=40051
>      scalar.dl.auditor.server.privileged_port=40052
>      ```

Reviewer 2: ```suggestion
1. **(ScalarDL Auditor mode only)** In the bastion server for ScalarDL Ledger, configure an existing kubeconfig file or add a new kubeconfig file to access the Kubernetes cluster for ScalarDL Auditor. For details on how to configure the kubeconfig file of each managed Kubernetes cluster, see [Configure kubeconfig](./CreateBastionServer.md#configure-kubeconfig).
2. Configure port forwarding to each service from the bastion server.
   * **ScalarDB Server**
     ```console
     kubectl port-forward -n <NAMESPACE> svc/<RELEASE_NAME>-envoy 60051:60051
     ```
   * **ScalarDL Ledger**
     ```console
     kubectl --context <CONTEXT_IN_KUBERNETES_FOR_SCALARDL_LEDGER> port-forward -n <NAMESPACE> svc/<RELEASE_NAME>-envoy 50051:50051
     kubectl --context <CONTEXT_IN_KUBERNETES_FOR_SCALARDL_LEDGER> port-forward -n <NAMESPACE> svc/<RELEASE_NAME>-envoy 50052:50052
     ```
   * **ScalarDL Auditor**
     ```console
     kubectl --context <CONTEXT_IN_KUBERNETES_FOR_SCALARDL_AUDITOR> port-forward -n <NAMESPACE> svc/<RELEASE_NAME>-envoy 40051:40051
     kubectl --context <CONTEXT_IN_KUBERNETES_FOR_SCALARDL_AUDITOR> port-forward -n <NAMESPACE> svc/<RELEASE_NAME>-envoy 40052:40052
     ```
3. Configure the properties file to access ScalarDB or ScalarDL via `localhost`.
   * **Client properties file for ScalarDB Server**
     ```properties
     scalar.db.contact_points=localhost
     scalar.db.contact_port=60051
     scalar.db.storage=grpc
     scalar.db.transaction_manager=grpc
     ```
   * **Client properties file for ScalarDL Ledger**
     ```properties
     scalar.dl.client.server.host=localhost
     scalar.dl.ledger.server.port=50051
     scalar.dl.ledger.server.privileged_port=50052
     ```
   * **Client properties file for ScalarDL Ledger with ScalarDL Auditor mode enabled**
     ```properties
     # Ledger
     scalar.dl.client.server.host=localhost
     scalar.dl.ledger.server.port=50051
     scalar.dl.ledger.server.privileged_port=50052
     
     # Auditor
     scalar.dl.client.auditor.enabled=true
     scalar.dl.client.auditor.host=localhost
     scalar.dl.auditor.server.port=40051
     scalar.dl.auditor.server.privileged_port=40052
     ```

```
